File: LeetCode/297. 二叉树的序列化与反序列化-DFS.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::VecDeque;
struct Codec {
	
}
const SEP: &str = ",";
const NULL: &str = "#";
/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl Codec {
    fn new() -> Self {
        Self {}
    }

    fn serialize(&self, root: Option<Rc<RefCell<TreeNode>>>) -> String {
        match root {
            None => NULL.to_string(),
            Some(node) => {
                let mut result: Vec<String> = vec![];
                let mut traversal: VecDeque<Option<Rc<RefCell<TreeNode>>>> = VecDeque::new();
                traversal.push_back(Some(node));
                while !traversal.is_empty() {
                    let cur = traversal.pop_front().unwrap();
                    match cur {
                        None => {
                            result.push(NULL.to_string());
                        },
                        Some(cur) => {
                            let cur = cur.borrow();
                            result.push(cur.val.to_string());
                            traversal.push_back(cur.left.clone());
                            traversal.push_back(cur.right.clone());
                        }
                    }
                    result.push(SEP.to_string());
                }
                result.pop();
                result.join("")
            }
        }
    }
	
    fn deserialize(&self, data: String) -> Option<Rc<RefCell<TreeNode>>> {
        let mut nodes: VecDeque<&str> = data.split(SEP).collect();
        let mut result: VecDeque<Rc<RefCell<TreeNode>>> = VecDeque::new();
        let root = nodes.pop_front();
        match root {
            None => None,
            Some(s) => {
                if s == NULL { None }
                else {
                    let root = TreeNode::new(s.parse().unwrap());
                    let root = Rc::new(RefCell::new(root));
                    result.push_back(root.clone());
                    while !result.is_empty() {
                        let mut cur = result.pop_front().unwrap();
                        let mut cur = cur.borrow_mut();
                        let leftVal = nodes.pop_front().unwrap();
                        if (leftVal != NULL) {
                            let left = Rc::new(RefCell::new(TreeNode::new(leftVal.to_string().parse().unwrap())));
                            result.push_back(left.clone());
                            cur.left = Some(left);
                        } else {
                            cur.left = None;
                        }
                        let rightVal = nodes.pop_front().unwrap();
                        if rightVal != NULL {
                            let right = Rc::new(RefCell::new(TreeNode::new(rightVal.to_string().parse().unwrap())));
                            result.push_back(right.clone());
                            cur.right = Some(right);
                        } else {
                            cur.right = None;
                        }
                    }
                    Some(root)
                }
            }
        }

    }
}

/**
 * Your Codec object will be instantiated and called as such:
 * let obj = Codec::new();
 * let data: String = obj.serialize(strs);
 * let ans: Option<Rc<RefCell<TreeNode>>> = obj.deserialize(data);
 */

File: LeetCode/297. 二叉树的序列化与反序列化-DFS.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from collections import deque
class Codec:
    SEP = ','
    NULL = '#'
    def serialize(self, root):
        if not root:
            return self.NULL
        q = deque([])
        result = []
        q.append(root)
        while len(q) != 0:
            cur = q.popleft()
            if cur:
                result.append(str(cur.val))
                result.append(self.SEP)
                q.append(cur.left)
                q.append(cur.right)
            else:
                result.append(self.NULL)
                result.append(self.SEP)
        result.pop()
        return ''.join(result)

    def deserialize(self, data):
        nodes = deque(data.split(self.SEP))
        if len(nodes) == 0:
            return None
        rootVal = nodes.popleft()
        if rootVal == self.NULL:
            return None
        root = TreeNode(rootVal)
        q = deque([])
        q.append(root)
        while len(q) != 0:
            cur = q.popleft()
            if len(nodes) != 0:
                leftVal = nodes.popleft()
                if leftVal != self.NULL:
                    cur.left = TreeNode(leftVal)
                    q.append(cur.left)
                else:
                    cur.left = None
            if len(nodes) != 0:
                rightVal = nodes.popleft()
                if rightVal != self.NULL:
                    cur.right = TreeNode(rightVal)
                    q.append(cur.right)
                else:
                    cur.right = None
        return root

        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))
